lib/upload-manager.js

```diff
@@ -0,0 +1,287 @@
+
+//@TODO: rename to something better
+/** Manager for uploading a file in chunks */
+class UploadManager extends EventEmitter {
+
+	/**
+	 * Create an upload manager
+	 * @param {BoxClient} client The client to use to upload the file
+	 * @param {Object} uploadSessionInfo The upload session info to use for chunked upload
+	 * @param {ReadableStream|Buffer|string} file The file to upload
+	 * @param {int} size The size of the file to be uploaded
+	 * @param {Object} [options] Optional parameters
+	 */
+	constructor(client, uploadSessionInfo, file, size, options) {
+
+		super();
+
+		this.client = client;
+		this.sessionID = uploadSessionInfo.upload_session_id;
+		this.partSize = uploadSessionInfo.part_size;
+		this.uploadSessionInfo = uploadSessionInfo;
+
+		if (file instanceof ReadableStream) {
+			// Pause the stream so we can read specific chunks from it
+			this.stream = file.pause();
+			this.streamBuffer = [];
+		} else if (file instanceof Buffer || typeof file === 'string') {
+			this.file = file;
+		} else {
+			throw new TypeError('file must be a Stream, Buffer, or string!');
+		}
+
+		this.size = size;
+		this.options = Object.assign({}, DEFAULT_OPTIONS, options);
+
+		this.isStarted = false;
+		this.numChunksInFlight = 0;
+		this.chunks = [];
+		this.position = 0;
+		this.fileHash = crypto.createHash('sha1');
+	}
+
+	/**
+	 * Start an upload
+	 * @returns {void}
+	 */
+	start() {
+
+		if (this.isStarted) {
+			return;
+		}
+
+		// Create the initial chunks
+		for (let i = 0; i < this.options.parallelism; i++) {
+			this._getNextChunk(chunk => (chunk ? this._uploadChunk(chunk) : this._commit()));
+		}
+		this.isStarted = true;
+	}
+
+	/**
+	 * Abort a running upload, which cancels all currently uploading chunks,
+	 * attempts to free up held memory, and aborts the upload session.  This
+	 * cannot be undone or resumed.
+	 * @returns {void}
+	 */
+	abort() {
+
+		this.chunks.forEach(chunk => chunk.removeAllListeners().cancel());
+		this.chunks = [];
+		this.file = null;
+		this.stream = null;
+
+		this.client.files.abortUploadSession(this.sessionID, err => {
+
+			if (err) {
+				this.emit('abortFailed', err);
+				return;
+			}
+
+			this.emit('aborted');
+		});
+	}
+
+	/**
+	 * Get the next chunk of the file to be uploaded
+	 * @param {Function} callback Called with the next chunk of the file to be uploaded
+	 * @returns {void}
+	 * @private
+	 */
+	_getNextChunk(callback) {
+
+		if (this.position >= this.size) {
+			callback(null);
+			return;
+		}
+
+		let buf;
+
+		if (this.file) {
+
+			// Buffer/string case, just get the slice we need
+			buf = this.file.slice(this.position, this.position + this.partSize);
+		} else if (this.streamBuffer.length) {
+
+			buf = this.streamBuffer.shift();
+		} else {
+
+			// Stream case, need to read
+			buf = this.stream.read(this.partSize);
+
+			if (!buf) {
+				// stream needs to read more, retry later
+				setImmediate(() => this._getNextChunk(callback));
```

Comment:
We don't need to wait a specific amount of time, but rather just until after the next round of I/O occurs, so setImmediate should allow us to retry at the soonest reasonable time.